core/src/main/scala/kafka/server/DelayedCreateTopics.scala

```diff
@@ -0,0 +1,94 @@
+  * limitations under the License.
+  */
+
+package kafka.server
+
+import kafka.api.LeaderAndIsr
+import kafka.common.TopicAndPartition
+import org.apache.kafka.common.protocol.Errors
+
+import scala.collection._
+
+/**
+  * The create metadata maintained by the delayed create operation
+  *
+  * TODO: local state doesn't count, need to know state of all relevant brokers
+  *
+  */
+case class CreateTopicMetadata(topic: String, replicaAssignments: Map[Int, Seq[Int]], error: Errors)
+
+/**
+  * A delayed create topics operation that can be created by the admin manager and watched
+  * in the topic purgatory
+  */
+class DelayedCreateTopics(delayMs: Long,
+                          createMetadata: Seq[CreateTopicMetadata],
+                          adminManager: AdminManager,
+                          responseCallback: Map[String, Errors] => Unit)
+  extends DelayedOperation(delayMs) {
+
+  /**
+    * The operation can be completed if all of the topics exist every partition has a leader in the controller.
+    * See KafkaController.onNewTopicCreation
+    */
+  override def tryComplete() : Boolean = {
+    trace(s"Trying to complete operation for $createMetadata")
+    trace(s"Live brokers ${adminManager.controller.controllerContext.liveBrokerIds}")
+
+    val leaderlessPartitionCount = createMetadata.foldLeft(0) { case (topicCounter, metadata) =>
+      topicCounter + missingLeaderCount(metadata.topic, metadata.replicaAssignments.keySet)
+    }
+
+    if (leaderlessPartitionCount == 0) {
+      trace("All partitions have a leader, completing the delayed operation")
+      forceComplete()
+    } else {
+      trace(s"$leaderlessPartitionCount partitions do not have a leader, not completing the delayed operation")
+      false
+    }
+  }
+
+  /**
+    * Check for partitions that are still missing a leader, update their error code and call the responseCallback
+    */
+  override def onComplete() {
+    trace(s"Completing operation for $createMetadata")
+    val results = createMetadata.map { metadata =>
+      // ignore topics that already have errors
+      if (metadata.error == Errors.NONE && missingLeaderCount(metadata.topic, metadata.replicaAssignments.keySet) > 0)
+        (metadata.topic, Errors.REQUEST_TIMED_OUT)
+      else
+        (metadata.topic, metadata.error)
+    }.toMap
+    responseCallback(results)
+  }
+
+  override def onExpiration(): Unit = { }
+
+  private def missingLeaderCount(topic: String, partitions: Set[Int]): Int = {
+    partitions.foldLeft(0) { case (counter, partition) =>
+      if (isMissingLeader(topic, partition)) counter + 1 else counter
+    }
+  }
+
+  private def isMissingLeader(topic: String, partition: Int): Boolean = {
+    val leadershipInfo = adminManager.controller.controllerContext.partitionLeadershipInfo
+    val leaderInfo =leadershipInfo.get(TopicAndPartition(topic, partition))
+    val partitionInfo = adminManager.metadataCache.getPartitionInfo(topic, partition)
```

Comment:
partitionInfo is unused. I thought we want to wait until the topic metadata is propagated to the controller. If so, we have to read from metadata cache since partitionLeadershipInfo only represents the state in the controller, but not metadata cache.